subsys/bluetooth/controller/ll_sw/ull_llcp.c

```diff
@@ -0,0 +1,470 @@
+struct mem_pool {
+	void *free;
+	u8_t *pool;
+};
+
+#define PROC_CTX_BUF_SIZE       WB_UP(sizeof(struct proc_ctx))
+#define TX_CTRL_BUF_SIZE        WB_UP(offsetof(struct node_tx, pdu) + offsetof(struct pdu_data, llctrl) + sizeof(struct pdu_data_llctrl))
+
+#if !defined(ULL_LLCP_UNITTEST)
+
+/* LLCP Allocations (normal) */
+
+static u8_t buffer_mem_tx[TX_CTRL_BUF_SIZE * 1];
+static struct mem_pool mem_tx = { .pool = buffer_mem_tx };
+static const size_t sz_mem_tx = TX_CTRL_BUF_SIZE;
+static const u32_t ncount_mem_tx = 1;
+
+static u8_t buffer_mem_ctx[TX_CTRL_BUF_SIZE * 1];
+static struct mem_pool mem_ctx = { .pool = buffer_mem_ctx };
+static const size_t sz_mem_ctx = PROC_CTX_BUF_SIZE;
+static const u32_t ncount_mem_ctx = 1;
+
+#else
+
+/* LLCP Allocations (unittest) */
+
+extern const size_t sz_mem_tx;
+extern const u32_t ncount_mem_tx;
+extern struct mem_pool mem_tx;
+
+extern const size_t sz_mem_ctx;
+extern const u32_t ncount_mem_ctx;
+extern struct mem_pool mem_ctx;
+
+#endif
+
+/*
+ * LLCP Resource Management
+ */
+
+static struct proc_ctx *proc_ctx_acquire(void)
+{
+	struct proc_ctx *ctx;
+
+	ctx = (struct proc_ctx *) mem_acquire(&mem_ctx.free);
+	return ctx;
+}
+
+static void proc_ctx_release(struct proc_ctx *ctx)
+{
+	mem_release(ctx, &mem_ctx.free);
+}
+
+static bool tx_alloc_peek(void)
+{
+	u16_t mem_free_count;
+
+	mem_free_count = mem_free_count_get(mem_tx.free);
+	return mem_free_count > 0;
+}
+
+static struct node_tx *tx_alloc(void)
+{
+	struct node_tx *tx;
+
+	tx = (struct node_tx *) mem_acquire(&mem_tx.free);
+	return tx;
+}
+
+/*
+ * Version Exchange Procedure Helper
+ */
+
+static void pdu_assemble_version_ind(struct pdu_data *pdu)
+{
+	u16_t cid;
+	u16_t svn;
+
+	pdu->ll_id = PDU_DATA_LLID_CTRL;
+	pdu->len = offsetof(struct pdu_data_llctrl, version_ind) + sizeof(struct pdu_data_llctrl_version_ind);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_VERSION_IND;
+	pdu->llctrl.version_ind.version_number = LL_VERSION_NUMBER;
+	cid = sys_cpu_to_le16(ll_settings_company_id());
+	svn = sys_cpu_to_le16(ll_settings_subversion_number());
+	pdu->llctrl.version_ind.company_id = cid;
+	pdu->llctrl.version_ind.sub_version_number = svn;
+}
+
+/*
+ * LLCP Local Procedure Common FSM
+ */
+
+static void lp_comm_tx(struct ull_cp_conn *conn, struct proc_ctx *ctx)
+{
+	struct node_tx *tx;
+	struct pdu_data *pdu;
+
+	/* Allocate tx node */
+	tx = tx_alloc();
+	LL_ASSERT(tx);
+
+	pdu = (struct pdu_data *)tx->pdu;
+
+	/* Assemble LL Control PDU */
+	switch (ctx->proc) {
+	case PROC_VERSION_EXCHANGE:
+		pdu_assemble_version_ind(pdu);
+		ctx->opcode = PDU_DATA_LLCTRL_TYPE_VERSION_IND;
+		break;
+	default:
+		/* Unknown procedure */
+		LL_ASSERT(0);
+	}
+
+	/* Enqueue LL Control PDU towards LLL */
+	ull_tx_q_enqueue_ctrl(conn->tx_q, tx);
+}
+
+
+static void lp_comm_st_idle(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt)
+{
+	switch (evt) {
+	case LP_COMMON_EVT_PREPARE:
+		if (!tx_alloc_peek() || ctx->pause || ctx->collision) {
```

Comment:
Why ctx->collision leads to WAIT_TX?

Reply:
??

Reply:
This is for the case where prior to the local procedure sending the first pdu, you already know that there is a remote procedure running that would give a collision, so WAIT_TX is wait state until the colliding procedure is finished.